[nnc] aotCompiler outputSpec support quantized outputs

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68151
* #68018
* __->__ #67711
* #67623

Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

IvanKobzarev committed:
[nnc] aotCompiler outputSpec support quantized outputs

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -82,6 +83,17 @@ std::unique_ptr<Function> compileMethod(
+    if (isQIntType(output.dtype_)) {
+      // Supporting only static qscale/qzero
+      output.qscale_ =
+          to<DoubleImm>(torch::jit::tensorexpr::IRSimplifier::simplify(
+                            ba.buf()->qscale()))
```
Should we simplify it whenever we create the buf, would it be cleaner (if it's possible at all)?

ZolotukhinM reviewed (approved):
LGTM!

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

IvanKobzarev committed:
Update on "[nnc] aotCompiler outputSpec support quantized outputs"


Differential Revision: [D32115833](https://our.internmc.facebook.com/intern/diff/D32115833)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32115833).

ljk53 commented on torch/csrc/jit/mobile/nnc/context.cpp:
```diff
@@ -48,17 +48,43 @@ OutputSpec::OutputSpec(const c10::IValue& value) {
       at::StringType::get(), at::AnyType::get());
   dict.insert("sizes", sizes_);
   dict.insert("dtype", dtype_);
+  if (qscale_) {
```
So looks like this won't break the backward compatibility for already compiled models.

cc: @priyaramani - as we discussed, I think we need to setup CIs to test already compiled models that we plan to run in production.
